### ISSUE TITLE
z-score values on box plots from PhenoSight analysis

### ISSUE THREAD

**Issue by Soorya19Pradeep** [czbiohub-sf/PhenoSight#28]
The z-score calculated on the box plots on the Analysis tab of PhenoSight, resulting from the analysis between infected and uninfected cells, doesn't seem to match the expectation. The value doesn't scale currently with the difference in significance of features between infected and uninfected cell populations. 

There was a discussion earlier to replace the z-score calculation with another scoring method. The z-scoring seems to be using the scipy [ranksum](https://github.com/czbiohub-sf/PhenoSight/blob/700e139d0f998eac9e8bc4e7abe5ffbe48662741/PhenoSight/main_widget.py#L1837) which might be assuming same size of cell population between the two cases. Did we decide to replace this with [Mann Whitney U test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html) which assumes two independent samples?

**Comment by nbathreya** [comment]
The z-score is calculated properly. Here are the steps:
1. For each condition, the data values are combined and ranked collectively. The lowest value gets assigned a rank of 1, the next lowest a rank of 2, and so on.
2. The sum of ranks within each group is calculated. This sum is used to compute the U statistic, which is one of the values used to calculate the z-score.
3. The z-score is then computed based on the U statistic, and it quantifies how much the distributions of ranks in the two groups differ from each other.

The only thing that might have to considered is the addition of normalizing each condition FOV by the number of cells in that FOV.

**Comment by Soorya19Pradeep** [comment]
Thanks @nbathreya! Does this mean the normalization has to be applied during the generation of the violin plot for the feature ranking as well as for different population sizes? Or is this already implemented?

**Comment by nbathreya** [comment]
Yes. The normalization has to be applied during the generation of the violin/box plot. It can avoid bias in the results. The normalization is done while classification and extraction of the top features. Here's the code snippet from evaluate_classifier function in main_widget.py:
# Now for remaining columns, replace np.nan with median of the column
        numerical_columns = df_filtered.select_dtypes(include=[np.number]).columns.tolist()
        df_filtered[numerical_columns] = df_filtered[numerical_columns].apply(lambda x: x.fillna(x.median()), axis=0)

        # Now try scaling again
        scaler = MinMaxScaler()
        # Create dataframe of scaled features only
        scaled_df = pd.DataFrame(scaler.fit_transform(df_filtered.drop('Label', axis=1)),
                                 columns=df_filtered.columns[1:],
                                 index=df_filtered.index)
        # Concatenate the 'Label' column with the scaled features
        df_scaled = pd.concat([df_filtered['Label'], scaled_df], axis=1)

        df_scaled.to_csv(os.path.join(self.parent_dir, 'normalized_data.csv'), index=False)

The only normalization step required is while plotting the violin and box plots.

**Comment by Soorya19Pradeep** [comment]
Thanks for the clarification @nbathreya! 

I used the cells with organelle LMAN1 to test this as I do not expect any changes in the morphology. But FOVs imaged have very different cell densities and show feature ranking based on organelle, membrane, and nuclear area as the top varying features. 

Can we add this as a modification implemented to the current version?

**Comment by nbathreya** [comment]
Were the z-scores shown on the plot normalized according to the number of cells in each FOV?

**Comment by Soorya19Pradeep** [comment]
Doesn't seem like it is. Here is a snapshot of a box plot of one of the top ranked feature:
![image](https://github.com/czbiohub-sf/PhenoSight/assets/101817974/0c00ebeb-bff0-46dc-962a-9413f8f94000)

Here is a snap of the list of features sorted based on rank:
![image](https://github.com/czbiohub-sf/PhenoSight/assets/101817974/3b237346-404a-4fa7-a1fd-9c44132df69c)

However, the points on the volcano plot corresponding to these top features are very close to (0,0). The points distant from (0,0) are other features like organelle compactness, organelle eccentricity, etc., which is what I expected for LMAN1.
![image](https://github.com/czbiohub-sf/PhenoSight/assets/101817974/de4761d1-15b0-40ba-95da-03aa824a1eac)

**Comment by Soorya19Pradeep** [comment]
I think I found one of the major reasons for this issue. 

The organelle segmentation fed into phenosight is a binary mask generated by AICS or Otsu segmentation. This binary mask is converted to labeled mask using `skimage.measure.label` function which assigns labels to different connected components in the label. These labels are used as child object labels as well. 

In the case of organelles like HSP90, the major part of the mask is extended over the entire field of view at high cell density. This makes one single huge label that extends over multiple cells. Whereas segmentation from residual background hot pixels makes even single-pixel labels. These segments are classified and stored in the CSV files and are equally significant in training the classifier. 

Concluding, the input to the classifier itself is not the right data for the classifier. I have summarized the same issue with images on [Confluence](https://czbiohub.atlassian.net/wiki/spaces/~6237962d1c7f6a007047e8ce/pages/3285614605/Analysis+on+PhenoSight+Insights) if interesting.
